src/socket/encrypter.rs

```diff
@@ -0,0 +1,250 @@
+use std::sync::Arc;
+
+use anyhow::Context;
+use bytes::Bytes;
+use earendil_crypt::{Fingerprint, IdentityPublic, IdentitySecret};
+use earendil_packet::crypt::{AeadKey, OnionPublic, OnionSecret};
+use replay_filter::ReplayFilter;
+use serde::{Deserialize, Serialize};
+use smol::{
+    channel::{Receiver, Sender},
+    future::FutureExt,
+    Task,
+};
+use stdcode::StdcodeSerializeExt;
+
+use crate::{daemon::context::DaemonContext, haven::HAVEN_FORWARD_DOCK};
+
+use super::{n2r_socket::N2rSocket, Endpoint, SocketRecvError, SocketSendError};
+
+#[derive(Clone)]
+pub struct Encrypter {
+    send_outgoing: Sender<Bytes>,
+    send_incoming: Sender<HavenMsg>,
+    _task: Arc<Task<anyhow::Result<()>>>,
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+pub enum HavenMsg {
+    ClientHs(Handshake),
+    ServerHs(Handshake),
+    Regular { nonce: u64, inner: Bytes },
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+pub struct Handshake {
+    id_pk: IdentityPublic,
+    eph_pk: OnionPublic,
+    sig: Bytes,
+}
+
+impl Encrypter {
+    pub fn client_new(
+        my_isk: IdentitySecret,
+        remote_ep: Endpoint,
+        rendezvous_fp: Option<Fingerprint>,
+        n2r_skt: N2rSocket,
+        send_incoming_decrypted: Sender<(Bytes, Endpoint)>,
+        ctx: DaemonContext,
+    ) -> Self {
+        let (send_out, recv_out) = smol::channel::bounded(1000);
+        let (send_in, recv_in) = smol::channel::bounded(1000);
+        let task = smolscale::spawn(enc_task(
+            my_isk,
+            n2r_skt,
+            remote_ep,
+            rendezvous_fp,
+            recv_in,
+            recv_out,
+            send_incoming_decrypted,
+            None,
+            ctx,
+        ));
+        Self {
+            send_outgoing: send_out,
+            send_incoming: send_in,
+            _task: Arc::new(task),
+        }
+    }
+
+    pub fn server_new(
+        my_isk: IdentitySecret,
+        remote_ep: Endpoint,
+        rendezvous_fp: Option<Fingerprint>,
+        n2r_skt: N2rSocket,
+        send_incoming_decrypted: Sender<(Bytes, Endpoint)>,
+        client_hs: Handshake,
+        ctx: DaemonContext,
+    ) -> Self {
+        let (send_out, recv_out) = smol::channel::bounded(1000);
```

Comment:
Considering we use pervasively `send` rather than `try_send` with these channels (so we do not need a large buffer to absorb bursts in case they cause loss), it's better to use a small buffer, like `1`, to not waste more space than necessary.

We can add more buffering later if we discover that it's necessary to, say, improve performance.